internal/registry/registry.go

```diff
@@ -0,0 +1,250 @@
+package registry
+
+import (
+	"bufio"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path/filepath"
+
+	ggcrname "github.com/google/go-containerregistry/pkg/name"
+	"github.com/pkg/errors"
+	"golang.org/x/mod/semver"
+	"gopkg.in/src-d/go-git.v4"
+
+	"github.com/buildpacks/pack/internal/buildpack"
+)
+
+const defaultRegistryURL = "https://github.com/buildpacks/registry-index"
+
+const defaultRegistryDir = "registry"
+
+type Buildpack struct {
+	Namespace string `json:"ns"`
+	Name      string `json:"name"`
+	Version   string `json:"version"`
+	Yanked    bool   `json:"yanked"`
+	Address   string `json:"addr"`
+}
+
+type Entry struct {
+	Buildpacks []Buildpack `json:"buildpacks"`
+}
+
+type Cache struct {
+	url  string
+	Root string
+}
+
+func NewRegistryCache(home, registryURL string) (Cache, error) {
+	if _, err := os.Stat(home); err != nil {
+		return Cache{}, err
+	}
+
+	normalizedURL, err := url.Parse(registryURL)
+	if err != nil {
+		return Cache{}, err
+	}
+
+	key := sha256.New()
+	key.Write([]byte(normalizedURL.String()))
+	cacheDir := fmt.Sprintf("%s-%s", defaultRegistryDir, hex.EncodeToString(key.Sum(nil)))
```

Comment:
@jkutner I might be misunderstanding how the registry cache works, but it appears every time we enter this logic flow we're creating a new unique cache.  Is this correct?  How will it find a previous registry cache with `hex.EncodeToString(key.Sum(nil))` resulting in a unique string?

Reply:
As long as the URL for the registry is the same, the registry directory will be the same. In this way, you can have multiple registry caches in your PACK_HOME at once (maybe a cache for heroku-registry, and a cache for buildpacks-io-registry or whatever)